lib/compress/zstd_compress.c

```diff
@@ -7022,6 +7092,246 @@ size_t ZSTD_compressSequences(ZSTD_CCtx* cctx,
+
+static size_t getNbSequencesFor1Block(const ZSTD_Sequence* seqs, size_t nbSeqs)
+{
+    size_t n;
+    assert(seqs);
+    for (n=0; n<nbSeqs; n++) {
+        if (seqs[n].matchLength == 0) {
+            assert(seqs[n].offset == 0);
+            return n+1;
+        }
+    }
+    RETURN_ERROR(externalSequences_invalid, "missing final block delimiter");
+}
+
+
+static size_t
+ZSTD_compressSequencesAndLiterals_internal(ZSTD_CCtx* cctx,
+                                void* dst, size_t dstCapacity,
+                          const ZSTD_Sequence* inSeqs, size_t nbSequences,
+                          const void* literals, size_t litSize)
+{
+    size_t cSize = 0;
+    BYTE* op = (BYTE*)dst;
+    int const repcodeResolution = (cctx->appliedParams.searchForExternalRepcodes == ZSTD_ps_enable);
+    assert(cctx->appliedParams.searchForExternalRepcodes != ZSTD_ps_auto);
+
+    DEBUGLOG(4, "ZSTD_compressSequencesAndLiterals_internal: nbSeqs=%zu, litSize=%zu", nbSequences, litSize);
+    RETURN_ERROR_IF(nbSequences == 0, externalSequences_invalid, "Requires at least 1 end-of-block");
+
+    /* Special case: empty frame */
+    if ((nbSequences == 1) && (inSeqs[0].litLength == 0)) {
+        U32 const cBlockHeader24 = 1 /* last block */ + (((U32)bt_raw)<<1);
+        RETURN_ERROR_IF(dstCapacity<4, dstSize_tooSmall, "No room for empty frame block header");
+        MEM_writeLE32(op, cBlockHeader24);
+        op += ZSTD_blockHeaderSize;
+        dstCapacity -= ZSTD_blockHeaderSize;
+        cSize += ZSTD_blockHeaderSize;
+    }
+
+    while (nbSequences) {
+        size_t compressedSeqsSize, cBlockSize, litConsumed;
+        size_t nbBlockSeqs = getNbSequencesFor1Block(inSeqs, nbSequences);
+        U32 const lastBlock = (nbBlockSeqs == nbSequences);
+        FORWARD_IF_ERROR(nbBlockSeqs, "Error while trying to determine nb of sequences for a block");
+        assert(nbBlockSeqs <= nbSequences);
+        ZSTD_resetSeqStore(&cctx->seqStore);
+
+        litConsumed = ZSTD_convertBlockSequences(cctx,
+                            inSeqs, nbBlockSeqs,
+                            repcodeResolution);
+        FORWARD_IF_ERROR(litConsumed, "Bad sequence conversion");
+        RETURN_ERROR_IF(litConsumed > litSize, externalSequences_invalid, "discrepancy between literals buffer and Sequences");
+        inSeqs += nbBlockSeqs;
+        nbSequences -= nbBlockSeqs;
+
+        /* Note: when blockSize is very small, other variant send it uncompressed.
+         * Here, we still send the sequences, because we don't have the original source to send it uncompressed.
+         * One could imagine it possible to reproduce the source from the sequences,
+         * but that's complex and costly memory intensive, which goes against the objectives of this variant. */
+
+        RETURN_ERROR_IF(dstCapacity < ZSTD_blockHeaderSize, dstSize_tooSmall, "not enough dstCapacity to write a new compressed block");
+
+        compressedSeqsSize = ZSTD_entropyCompressSeqStore_internal(
```

Comment:
Do we need to call `ZSTD_entropyCompressSeqStore_internal()` or can we call `ZSTD_entropyCompressSeqStore_wExtLitBuffer()`?

Reply:
The initial code was invoking `ZSTD_entropyCompressSeqStore_wExtLitBuffer()`.
But because `ZSTD_entropyCompressSeqStore_wExtLitBuffer()` is essentially `ZSTD_entropyCompressSeqStore()`, just with a selectable `litBuffer`,
it adds a few constraints, which favor the creation of uncompressed blocks.
In the normal case, it's fine (an uncompressed block is preferable to a compressed one with only 1 byte of saving),
but in the case of `ZSTD_compressSequencesAndLiterals()`, it is catastrophic, because this variant is unable to generate an uncompressed block (since it doesn't see the original data). This would result in an error.
So we prefer skipping these additional checks, by invoking the next stage directly, `ZSTD_entropyCompressSeqStore_internal()`.